File: workspaces/telepath-js/source/secure-channel.js
import base64url from 'base64url'
import { random, encrypt, decrypt, nonceSize } from '@cogitojs/crypto'
import { Poller } from './poller'

class SecureChannel {
  constructor ({ queuing, id, key, appName, socketIOService }) {
    this.queuing = queuing
    this.socketIOService = socketIOService
    this.id = id
    this.key = key
    this.appName = appName
    this.poller = new Poller({
      pollFunction: () => queuing.receive(`${id}.blue`),
      interval: 1000,
      retries: 600
    })
  }

  async startNotifications (notificationHandler, errorHandler) {
    this.notificationHandler = notificationHandler
    this.notificationErrorHandler = errorHandler
    await this.socketIOService.start(
      this.id,
      this.onEncryptedNotification.bind(this),
      errorHandler
    )
  }

  async send (message) {
    const queueId = `${this.id}.red`
    const nonceAndCypherText = await this.encrypt(message)
    await this.queuing.send(queueId, nonceAndCypherText)
  }

  async receive () {
    const received = await this.poller.poll()
    if (!received) {
      return null
    }
    return this.decrypt(received)
  }

  async notify (message) {
    const nonceAndCypherText = await this.encrypt(message)
    this.socketIOService.notify(nonceAndCypherText)
  }

  async onEncryptedNotification (data) {
    const message = await this.decrypt(data)
    this.notificationHandler(message)
  }

  async encrypt (message) {
    const nonce = await random(await nonceSize())
    const plainText = new Uint8Array(Buffer.from(message))
    const cypherText = await encrypt(plainText, nonce, this.key)
    return Buffer.concat([Buffer.from(nonce), Buffer.from(cypherText)])
  }

  async decrypt (data) {
    const nonceAndCypherText = new Uint8Array(data)
    const nonce = nonceAndCypherText.slice(0, await nonceSize())
    const cypherText = nonceAndCypherText.slice(await nonceSize())
    return decrypt(cypherText, nonce, this.key, 'text')
  }

  createConnectUrl (baseUrl) {
    const encodedKey = base64url.encode(this.key)
    const encodedAppName = base64url.encode(this.appName)
    return `${baseUrl}/telepath/connect#I=${
      this.id
    }&E=${encodedKey}&A=${encodedAppName}`
  }
}

export { SecureChannel }

File: workspaces/telepath-ios/Telepath/SecureChannel.swift
import Sodium
import SocketIO

public struct SecureChannel {
    public let id: ChannelID
    public let appName: String

    var queuing: QueuingService
    let socketIOService: SocketIOService
    let key: ChannelKey
    let receivingQueue: QueueID
    let sendingQueue: QueueID
    let notificationHandler: NotificationHandler?
    var notificationsStarted: Bool { return socketIOService.started }

    init(queuing: QueuingService, socketIOService: SocketIOService,
         notificationHandler: NotificationHandler?,
         id: ChannelID, key: ChannelKey, appName: String) {
        self.queuing = queuing
        self.socketIOService = socketIOService
        self.notificationHandler = notificationHandler
        self.key = key
        self.id = id
        self.appName = appName
        self.receivingQueue = id + ".red"
        self.sendingQueue = id + ".blue"
    }

    public mutating func invalidate() {
        queuing.invalidate()
    }

    public func send(message: String, completion: @escaping (Error?) -> Void) {
        let plainText = message.data(using: .utf8)!
        guard let cypherText = key.encrypt(plainText: plainText) else {
            completion(Failure.invalidKey)
            return
        }
        queuing.send(queueId: sendingQueue, message: cypherText) { error in
            guard error == nil else {
                completion(Failure.sendingFailed(cause: error!))
                return
            }
            completion(nil)
        }
    }

    public func receive(completion: @escaping (String?, Error?) -> Void) {
        queuing.receive(queueId: receivingQueue) { data, error in
            guard error == nil else {
                completion(nil, Failure.receivingFailed(cause: error!))
                return
            }
            guard let cypherText = data else {
                completion(nil, nil)
                return
            }
            guard let plainText = try? self.key.decrypt(cypherText: cypherText) else {
                completion(nil, Failure.decryptionFailed)
                return
            }
            completion(String(data: plainText, encoding: .utf8), nil)
        }
    }

    public func startNotifications(completion: CompletionHandler?) {
        socketIOService.start(channelID: id,
                              onNotification: onEncryptedNotification,
                              onError: notificationHandler?.on(error:),
                              completion: completion)
    }

    public func notify(message: String) {
        let plainText = message.data(using: .utf8)!
        if let cypherText = key.encrypt(plainText: plainText) {
            socketIOService.notify(data: cypherText)
        }
    }

    func onEncryptedNotification(data: Data) {
        guard let plainText = try? self.key.decrypt(cypherText: data),
              let message = String(data: plainText, encoding: .utf8) else {
            return
        }
        notificationHandler?.on(notification: message)
    }

    enum Failure: Error {
        case invalidKey
        case sendingFailed (cause: Error)
        case receivingFailed (cause: Error)
        case decryptionFailed
    }
}

extension Array where Element == UInt8 {
    func encrypt(plainText: Data) -> Data? {
        let box = Sodium().secretBox
        guard let sealed: [UInt8] = box.seal(message: [UInt8](plainText),
                                             secretKey: self) else {
            return nil
        }
        return Data(sealed)
    }

    func decrypt(cypherText: Data) throws -> Data {
        let box = Sodium().secretBox
        let opened = box.open(nonceAndAuthenticatedCipherText: [UInt8](cypherText), secretKey: self)
        guard let result = opened else {
            throw Errors.decryptionFailed
        }
        return Data(result)
    }

    enum Errors: Error {
        case decryptionFailed
    }
}
